command/alloc_signal.go

```diff
@@ -0,0 +1,192 @@
+	var verbose bool
+	var signal string
+
+	flags := c.Meta.FlagSet(c.Name(), FlagSetClient)
+	flags.Usage = func() { c.Ui.Output(c.Help()) }
+	flags.BoolVar(&verbose, "verbose", false, "")
+	flags.StringVar(&signal, "s", "SIGKILL", "")
+
+	if err := flags.Parse(args); err != nil {
+		return 1
+	}
+
+	// Check that we got exactly one alloc
+	args = flags.Args()
+	if len(args) < 1 || len(args) > 2 {
+		c.Ui.Error("This command takes up to two arguments: <alloc-id> <task>")
+		c.Ui.Error(commandErrorText(c))
+		return 1
+	}
+
+	allocID := args[0]
+
+	// Truncate the id unless full length is requested
+	length := shortId
+	if verbose {
+		length = fullId
+	}
+
+	// Query the allocation info
+	if len(allocID) == 1 {
+		c.Ui.Error(fmt.Sprintf("Alloc ID must contain at least two characters."))
+		return 1
+	}
+
+	allocID = sanitizeUUIDPrefix(allocID)
+
+	// Get the HTTP client
+	client, err := c.Meta.Client()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error initializing client: %s", err))
+		return 1
+	}
+
+	allocs, _, err := client.Allocations().PrefixList(allocID)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error querying allocation: %v", err))
+		return 1
+	}
+
+	if len(allocs) == 0 {
+		c.Ui.Error(fmt.Sprintf("No allocation(s) with prefix or id %q found", allocID))
+		return 1
+	}
+
+	if len(allocs) > 1 {
+		// Format the allocs
+		out := formatAllocListStubs(allocs, verbose, length)
+		c.Ui.Error(fmt.Sprintf("Prefix matched multiple allocations\n\n%s", out))
+		return 1
+	}
+
+	// Prefix lookup matched a single allocation
+	alloc, _, err := client.Allocations().Info(allocs[0].ID, nil)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error querying allocation: %s", err))
+		return 1
+	}
+
+	var taskName string
+	if len(args) == 2 {
+		// Validate Task
+		taskName = args[1]
+		err := validateTaskExistsInAllocation(taskName, alloc)
+		if err != nil {
+			c.Ui.Error(err.Error())
+			return 1
+		}
+	}
+
+	err = client.Allocations().Signal(alloc, nil, taskName, signal)
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("Error signalling allocation: %s", err))
+		return 1
+	}
+
+	return 0
+}
+
+func (a *AllocSignalCommand) Synopsis() string {
+	return "Signal a running allocation"
+}
+
+func (c *AllocSignalCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-s":       complete.PredictNothing,
+			"-verbose": complete.PredictNothing,
+		})
+}
+func (c *AllocSignalCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictFunc(func(a complete.Args) []string {
+		client, err := c.Meta.Client()
+		if err != nil {
+			return nil
+		}
+
+		count := len(a.All)
+		if count > 3 {
+			// Too many args, do nothing, we only take an allocation and single task
+			return []string{}
+		} else if count <= 2 {
+			// Searching for an alloc because we either have only [signal] or [signal prefix]
+			resp, _, err := client.Search().PrefixSearch(a.Last, contexts.Allocs, nil)
```

Comment:
this will return all extant allocations, even if they are stopped... do we want to (can we afford to) filter down by running allocations?

Reply:
there's no good way to do this here tbh - I assume people usually want this for "i know the first few characters", which there's usually fairly minimal overlap of